uccb/src/bin/validator-oracle.rs

```diff
@@ -0,0 +1,261 @@
+    pub async fn start(&mut self) -> Result<()> {
+        info!(
+            "Starting validator oracle with signer address {}",
+            self.signer.address()
+        );
+
+        let validators = self.get_stakers().await?;
+        info!("Current validator set is: {validators:?}");
+
+        let (sender, receiver) = watch::channel(validators);
+
+        let mut handles = vec![];
+        for chain_client in &self.chain_clients {
+            let mut receiver = receiver.clone();
+            let chain_client = chain_client.clone();
+            let validator_manager_abi = self.validator_manager_abi.clone();
+            let handle = tokio::spawn(async move {
+                loop {
+                    let validators = receiver.borrow_and_update().clone();
+                    if let Err(e) = Self::update_validator_manager(
+                        &chain_client,
+                        &validators,
+                        validator_manager_abi.clone(),
+                    )
+                    .await
+                    {
+                        error!(
+                            "Failed updating the validator manager on {}: {e}",
+                            chain_client.rpc_url
+                        );
+                    }
+
+                    if receiver.changed().await.is_err() {
+                        break;
+                    }
+                }
+            });
+            handles.push(handle);
+        }
+
+        let result = self.listen_to_staker_updates(sender).await;
+
+        for handle in handles {
+            handle.await?;
+        }
+
+        result
+    }
+
+    async fn listen_to_staker_updates(
+        &mut self,
+        sender: watch::Sender<Vec<Address>>,
+    ) -> Result<()> {
+        let signatures: Vec<FixedBytes<32>> = vec![
+            contracts::deposit::STAKER_ADDED_EVT.signature(),
+            contracts::deposit::STAKER_REMOVED_EVT.signature(),
+        ]
+        .into_iter()
+        .map(|x| FixedBytes::<32>::from(x.as_fixed_bytes()))
+        .collect();
+
+        let filter = Filter::new()
+            .address(contract_addr::DEPOSIT)
+            .event_signature(signatures)
+            .from_block(BlockNumberOrTag::Finalized);
+
+        let subscription = self
+            .zq2_chain_client()
+            .provider
+            .subscribe_logs(&filter)
+            .await?;
+        let mut stream = subscription.into_stream();
+        while let Some(log) = stream.next().await {
+            info!("Received validator update: {log:?}");
+
+            //let block_number = self.zq2_chain_client().provider.get_block_number().await?;
+            let block_number: u64 = self
+                .zq2_chain_client()
+                .provider
+                .raw_request("eth_blockNumber".into(), (BlockNumberOrTag::Finalized,))
+                .await?;
+
+            if let Some(log_block_number) = log.block_number {
+                if log_block_number >= block_number {
```

Comment:
I think this check would always be `false`, since the subscription will always return logs from the latest block and the finalized block will always be less than this.

Instead I think we need to keep received logs 'pending' until the block they were emitted in is finalized. And also handle the case where a log is dropped because a different block at that height was finalized, which doesn't contain the log.

Reply:
Imo the check is always true since `log_block_number` is the latest block number while `block_number` is the finalized block number which is always smaller than the latest one. Otherwise I'd implement it as you suggested @JamesHinshelwood

Reply:
Yes sorry, the check is always true as you say :)